mypy/checkexpr.py

```diff
@@ -1350,91 +1347,56 @@ def overload_erased_call_targets(self,
+            return res
+        first_union = next(typ for typ in arg_types if self.real_union(typ))
+        idx = arg_types.index(first_union)
+        assert isinstance(first_union, UnionType)
+        returns = []
+        inferred_types = []
+        for item in first_union.relevant_items():
+            new_arg_types = arg_types.copy()
+            new_arg_types[idx] = item
+            sub_result = self.union_overload_result(plausible_targets, args, new_arg_types,
+                                                    arg_kinds, arg_names, callable_name,
+                                                    object_type, context, arg_messages)
+            if sub_result is not None:
+                ret, inferred = sub_result
+                returns.append(ret)
+                inferred_types.append(inferred)
+            else:
                 return None
-            for i, (new_kind, target_kind) in enumerate(zip(new_kinds, target.arg_kinds)):
-                if new_kind == target_kind:
-                    continue
-                elif new_kind in (ARG_POS, ARG_OPT) and target_kind in (ARG_POS, ARG_OPT):
-                    new_kinds[i] = ARG_POS
-                else:
-                    return None
 
-            for i, arg in enumerate(target.arg_types):
-                new_args[i].append(arg)
-            new_returns.append(target.ret_type)
-
-        union_count = 0
-        final_args = []
-        for args_list in new_args:
-            new_type = UnionType.make_simplified_union(args_list)
-            union_count += 1 if isinstance(new_type, UnionType) else 0
-            final_args.append(new_type)
-
-        # TODO: Modify this check to be less conservative.
-        #
-        # Currently, we permit only one union in the arguments because if we allow
-        # multiple, we can't always guarantee the synthesized callable will be correct.
-        #
-        # For example, suppose we had the following two overloads:
-        #
-        #     @overload
-        #     def f(x: A, y: B) -> None: ...
-        #     @overload
-        #     def f(x: B, y: A) -> None: ...
-        #
-        # If we continued and synthesize "def f(x: Union[A,B], y: Union[A,B]) -> None: ...",
-        # then we'd incorrectly accept calls like "f(A(), A())" when they really ought to
-        # be rejected.
-        #
-        # However, that means we'll also give up if the original overloads contained
-        # any unions. This is likely unnecessary -- we only really need to give up if
-        # there are more then one *synthesized* union arguments.
-        if union_count >= 2:
-            return None
+        if returns:
+            print('un', returns, UnionType.make_simplified_union(returns, context.line, context.column))
+            return (UnionType.make_simplified_union(returns, context.line, context.column),
+                    UnionType.make_simplified_union(inferred_types, context.line, context.column))
```

Comment:
I am not sure we can use sets (although I really want to), the problem is that we need to have a stable/predictable order of union items to avoid flaky tests.

Reply:
OK, I think I figured out how to avoid multiple intermediate unions, and still have stable order, we can return a set of pairs (indexes of matched overload, return type), at the very end we construct a union once but before sort items by index.

Another important note: I think returning a union of inferred callable types may be wrong here, we should actually return a single unioned callable (or fall back to `Callable[..., Union[<ret_types>]]` if the matched variants are too different to combine). The motivation is that `(Union[int -> None, str -> None])(Union[int, str])` is not a valid call, maybe some code (even future code) relies on the fact that inferred callable type and initial arg types are consistent.